jwst/saturation/saturation.py

```diff
@@ -13,89 +13,103 @@
-    file.
+    Apply saturation flagging based on threshold values stored in the
+    saturation reference file.
 
     Parameters
     ----------
-    input_model: data model object
+    input_model : `~jwst.datamodels.RampModel`
         The input science data to be corrected
 
-    ref_model: data model object
-        Saturation reference file mode object
+    ref_model : `~jwst.datamodels.SaturationModel`
+        Saturation reference file data model
 
     Returns
     -------
-    output_model: data model object
-        object having GROUPDQ array saturation flags set
+    output_model : `~jwst.datamodels.RampModel`
+        Data model with saturation flags set in GROUPDQ array
     """
 
-    ramparr = input_model.data
-    # Was IRS2 readout used?
+    ramp_array = input_model.data
+    nints = ramp_array.shape[0]
+    ngroups = ramp_array.shape[1]
+    sat_flag = dqflags.group['SATURATED']
+    detector = input_model.meta.instrument.detector
+
+    # If NIRSpec IRS2 readout mode was used, create a mask
+    # of the appropriate size
     is_irs2_format = pipe_utils.is_irs2(input_model)
     if is_irs2_format:
         irs2_mask = x_irs2.make_mask(input_model)
 
-   # Create the output model as a copy of the input
+    # Create the output model as a copy of the input
     output_model = input_model.copy()
     groupdq = output_model.groupdq
 
-    # Extract subarray from reference file, if necessary
+    # Extract subarray from saturation reference file, if necessary
     if reffile_utils.ref_matches_sci(input_model, ref_model):
-        satmask = ref_model.data
-        dqmask = ref_model.dq
+        sat_thresh = ref_model.data
+        sat_dq = ref_model.dq
     else:
         log.info('Extracting reference file subarray to match science data')
         ref_sub_model = reffile_utils.get_subarray_model(input_model, ref_model)
-        satmask = ref_sub_model.data.copy()
-        dqmask = ref_sub_model.dq.copy()
+        sat_thresh = ref_sub_model.data.copy()
+        sat_dq = ref_sub_model.dq.copy()
         ref_sub_model.close()
 
-    # For pixels flagged in reference file as NO_SAT_CHECK, set the dq mask
-    #   and saturation mask
-    wh_sat = np.bitwise_and(dqmask, dqflags.pixel['NO_SAT_CHECK'])
-    dqmask[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = dqflags.pixel['NO_SAT_CHECK']
-    satmask[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = HUGE_NUM
-    # Correct saturation values for NaNs in the ref file
-    correct_for_NaN(satmask, dqmask)
-
-    dq_flag = dqflags.group['SATURATED']
+    # For pixels flagged in reference file as NO_SAT_CHECK,
+    # set the output dq mask to NO_SAT_CHECK and set the
+    # saturation check threshold to the A-to-D converter limit.
+    wh_sat = np.bitwise_and(sat_dq, dqflags.pixel['NO_SAT_CHECK'])
+    sat_dq[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = dqflags.pixel['NO_SAT_CHECK']
+    sat_thresh[wh_sat == dqflags.pixel['NO_SAT_CHECK']] = ATOD_LIMIT
+    del wh_sat
```

Comment:
Looking at the original code, it first found all the places in the ref file DQ array that have NO_SAT_CHECK and then set all those pixels equal to NO_SAT_CHECK. So it wasn't actually making any change to the NO_SAT_CHECK values themselves, but instead was wiping out any other DQ values that might've been set for those pixels. These values eventually then get propagated into the science data PIXELDQ array. I think we should be propagating *all* DQ values from the ref file, so I'm modifying this logic in the way you suggest, by simply removing line 71, because there's no reason to reset values in the array that were used to find the NO_SAT_CHECK flags in the first place.